src/shims/unix/socket.rs

```diff
@@ -20,17 +38,121 @@ impl FileDescription for SocketPair {
         self: Box<Self>,
         _communicate_allowed: bool,
     ) -> InterpResult<'tcx, io::Result<()>> {
+        // This is used to signal socketfd of other side that there is no writer to it's readbuf.
+        // If the upgrade fails, there is no need to update as all read ends has been dropped.
+        if let Some(writebuf) = self.writebuf.upgrade() {
+            writebuf.borrow_mut().buf_has_writer = false;
+        };
         Ok(Ok(()))
     }
+
+    fn read<'tcx>(
+        &mut self,
+        _communicate_allowed: bool,
+        bytes: &mut [u8],
+        ecx: &mut MiriInterpCx<'tcx>,
+    ) -> InterpResult<'tcx, io::Result<usize>> {
+        let request_byte_size = bytes.len();
+        let data_size: usize = self.readbuf.borrow().buf.len();
+        let readbuf_has_writer = self.readbuf.borrow().buf_has_writer;
+
+        // Always succeed on read size 0.
+        if request_byte_size == 0 {
+            return Ok(Ok(0));
+        }
+
+        if self.is_nonblock {
+            if !readbuf_has_writer {
+                if data_size == 0 {
+                    // Non-blocking socketpair with no writer and empty buffer.
+                    return Ok(Ok(0));
+                }
+            } else if data_size == 0 {
+                // Non-blocking socketpair with writer and empty buffer.
+                // https://linux.die.net/man/2/read
+                // EAGAIN or EWOULDBLOCK can be returned for socket,
+                // POSIX.1-2001 allows either error to be returned for this case.
+                // Since there is no ErrorKind for EAGAIN, WouldBlock is used.
+                return Ok(Err(Error::from(ErrorKind::WouldBlock)));
+            }
+        } else {
+            if !readbuf_has_writer {
+                if data_size == 0 {
+                    // Blocking socketpair with no writer and empty buffer.
+                    return Ok(Ok(0));
+                }
+            } else if data_size == 0 {
+                // Blocking socketpair with writer and empty buffer.
+                // FIXME: blocking is currently not supported
+                throw_unsup_format!("socketpair read: blocking isn't supported yet");
+            }
+        }
```

Comment:
you can deduplicate these arms by moving the `self.is_nonblock` check into the `data_size == 0` arm